crypto/params.c

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright (c) 2019, Oracle and/or its affiliates.  All rights reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/bn.h>
+#include <openssl/params.h>
+#include <openssl/err.h>
+#include <openssl/cryptoerr.h>
+
+OSSL_PARAM *OSSL_PARAM_locate(const OSSL_PARAM *p, const char *key)
+{
+    for (; p->key != NULL; p++)
+        if (strcmp(key, p->key) == 0)
+            return (OSSL_PARAM *)p;
+    return NULL;
+}
+
+static size_t OSSL_PARAM_get_width(int type)
+{
+    switch (type) {
+    case OSSL_PARAM_INT:        return sizeof(int);
+    case OSSL_PARAM_UINT:       return sizeof(unsigned int);
+    case OSSL_PARAM_INT64:      return sizeof(int64_t);
+    case OSSL_PARAM_UINT64:     return sizeof(uint64_t);
+    case OSSL_PARAM_LONG:       return sizeof(long);
+    case OSSL_PARAM_ULONG:      return sizeof(unsigned long);
+    case OSSL_PARAM_SIZET:      return sizeof(size_t);
+    case OSSL_PARAM_DOUBLE:     return sizeof(double);
+    case OSSL_PARAM_POINTER:    return sizeof(void *);
+    }
+    return -1;
+}
+
+static int OSSL_PARAM_get_int_common(const OSSL_PARAM *p, const char *key,
+                                     void *val, unsigned long type)
+{
+    size_t width;
+
+    if ((p = OSSL_PARAM_locate(p, key)) == NULL) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_GET_INT_COMMON, CRYPTO_R_PARAM_NOT_FOUND);
+        ERR_add_error_data(2, "param name=", key);
+        return 0;
+    }
+
+    /* Type safety. */
+    width = OSSL_PARAM_get_width(type);
+    if (p->data_type != type || p->buffer_size != width) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_GET_INT_COMMON, CRYPTO_R_TYPE_MISMATCH);
+        return 0;
+    }
+
+    memcpy(val, p->buffer, width);
```

Comment:
I can't say I'm massively keen on memcpy for assignment

Reply:
Because this is a generic function, there's no way around it.  The other option would be to implement it as a macro, but I've seen that you're not a fan of that either :wink:

Reply:
Well there are other options:

````
switch(type) {
case OSSL_PARAM_INT64:
    *(uint64_t *)val = *(uint64_t *)p->buffer;
    break;
/* etc */
....
}
````

Reply:
Arguably rather than having a single "buffer" type it should be a union for all the various underlying native C types we want to support.

Reply:
It would be a union pointers to different types then...  and well, we have those a little here and there, and I'm frankly not sure how much good it does for us...

Reply:
I trust that the compiler can use intrinsics to make the generic memcpy fast.  People time is more important than runtime for this function.

Reply:
We cannot have a union because you can't initialize specific arms of the union until C99.

Reply:
It is "well-known" that modern copies elide the memcpy when appropriate for the architecture in question.

W.r.t. C99 union initializers, "does MSVC support it?  What about VMS?"  (This should surprise no one, but my stance is that we should not get held up on slavish adherence to outdated standards when practical.)